### PR TITLE
Only request the necessary document and project in GraphBuilder.PopulateMapsForFileInputNode

### PR DESCRIPTION
**Pull request by ToddGrun**
This method previously enumerated documents and projects until it found ones with the appropriate path, creating many unused documents and projects along the way.

Instead, as the file path informtion is stored in both the projectstate and documentstate, we can walk those state objects and find the appropriate ProjectId and DocumentId. Once we have those, it's a direct shot to get the Project and Document.

Profiles were gathered against typing one 50 character line of text slowly (1 sec per char) in a large .cs file (LanguageParser.cs in Roslyn sln) with C# LSP turned on.

Old allocation:

image

New allocation:

image

### CODE REVIEW

File: src/VisualStudio/Core/Def/Progression/GraphBuilder.cs [hunk 1]
```diff
@@ -76,22 +76,26 @@ private void PopulateMapsForFileInputNode(GraphNode inputNode, CancellationToken
                     return;
                 }
 
-                var project = _solution.Projects.FirstOrDefault(
-                    p => string.Equals(p.FilePath, projectPath.OriginalString, StringComparison.OrdinalIgnoreCase));
-                if (project == null)
+                var projectState = _solution.SolutionState.ProjectStates.FirstOrDefault(
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Progression/GraphBuilder.cs:79]: consider doc'ing.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/VisualStudio/Core/Def/Progression/GraphBuilder.cs [hunk 2]
```diff
@@ -76,22 +76,30 @@ private void PopulateMapsForFileInputNode(GraphNode inputNode, CancellationToken
...
                 _nodeToContextProjectMap.Add(inputNode, project);
 
-                var document = project.Documents.FirstOrDefault(
-                    d => string.Equals(d.FilePath, filePath.OriginalString, StringComparison.OrdinalIgnoreCase));
-                if (document == null)
+                // Don't walk project.Documents as that will possibly create documents that don't match
+                // our criteria. Instead, walk the document states to find the appropriate document id.
+                var documentState = project.State.DocumentStates.States.FirstOrDefault(
+                    d => string.Equals(d.Value.FilePath, filePath.OriginalString, StringComparison.OrdinalIgnoreCase));
+                if (documentState.Key == null)
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Progression/GraphBuilder.cs:96]: What might be better still: just call [GetDocumentIdsWithFilePath(String)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.solution.getdocumentidswithfilepath?view=roslyn-dotnet-4.7.0#microsoft-codeanalysis-solution-getdocumentidswithfilepath(system-string)), and if we get more than one then compare the project's paths. That way we'll significantly cut down on string comparisons.

**Review by jasonmalinowski — COMMENTED** [review]
The change is good, but I think we can do a lot better?

File: src/VisualStudio/Core/Def/Progression/GraphBuilder.cs (hunk 2 above)
Comment (ToddGrun) [reply src/VisualStudio/Core/Def/Progression/GraphBuilder.cs:96]: Ooh, I like that. Let me give that a whirl...

Comment (ToddGrun) [reply src/VisualStudio/Core/Def/Progression/GraphBuilder.cs]: Yeah, much nicer. Cleaner and more efficient!

Comment (ToddGrun) [reply src/VisualStudio/Core/Def/Progression/GraphBuilder.cs]: test is failing, not obvious why. I'll dig in more in the morning.

**Comment by ToddGrun** [comment]
@jasonmalinowski -- Can you take another look to make sure this matches your expectation? Thanks!

**Review by jasonmalinowski — CHANGES_REQUESTED** [review]
One small but critical bug. I wish GitHub had a "you really have to fix this bit, but otherwise approved".

File: src/VisualStudio/Core/Def/Progression/GraphBuilder.cs [hunk 3]
```diff
@@ -76,22 +76,28 @@ private void PopulateMapsForFileInputNode(GraphNode inputNode, CancellationToken
...
-                if (project == null)
+                var docIdsWithPath = _solution.GetDocumentIdsWithFilePath(filePath.OriginalString);
+                Document? document = null;
+                Project? project = null;
+
+                foreach (var docIdWithPath in docIdsWithPath)
                 {
-                    return;
+                    var projectState = _solution.GetProjectState(docIdWithPath.ProjectId);
+                    if (projectState != null &&
```
[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/VisualStudio/Core/Def/Progression/GraphBuilder.cs]: @jasonmalinowski -- all good now?

**Review by jasonmalinowski — APPROVED** [review]
:shipit:

**Pull request merged**